core/uploader.go

```diff
@@ -0,0 +1,65 @@
+package core
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"time"
+
+	"github.com/livepeer/go-tools/drivers"
+)
+
+func Upload(input io.Reader, outputURI string, waitBetweenWrites, writeTimeout time.Duration) error {
+	storageDriver, err := drivers.ParseOSURL(outputURI, true)
+	if err != nil {
+		return err
+	}
+	session := storageDriver.NewSession("")
+	if err != nil {
+		return err
+	}
+
+	var fileContents = []byte{}
+	var lastWrite = time.Now()
+
+	scanner := bufio.NewScanner(input)
+
+	// We have to use a custom scanner because the default one is designed for text and will
+	// split on and drop newline characters
+	scanner.Split(func(data []byte, atEOF bool) (advance int, token []byte, err error) {
+		// If we have reached the end of the input, return 0 bytes and no error.
+		if atEOF {
+			return 0, nil, nil
+		}
+
+		// Read the entire input as one line by advancing the buffer to its end.
+		return len(data), data, nil
+	})
+
+	for scanner.Scan() {
+		b := scanner.Bytes()
+		fileContents = append(fileContents, b...)
+
+		// Only write the latest version of the data that's been piped in if enough time has elapsed since the last write
+		if lastWrite.Add(waitBetweenWrites).Before(time.Now()) {
```

Comment:
IIUC this means that we will only write to storage when a new write comes after the specified delay. Assuming a waitBetweenWrites of 5s a write pattern of
 1. write 100 bytes
 2. after 1s write 100 bytes
 3. after 1s write 100 bytes
 4. sleep for 1 hour
 5. write 100 bytes

Would only write during step 1 and then step 5 with the combination of 2, 3 and 5 contents. Is that correct?

I don't really know what's the Mist protocol there, but we might want to consider using a [`time.Timer`](https://pkg.go.dev/time#Timer.Reset) instead (or ticker idk), which is reset after every new write to only write after `waitBetweenWrites` (yeah it'd need a separate goroutine to do the writes 😬). In that case, in the example above, the writes would happen `5s` after the step 3, then `5s` after step 5. 

WDYT?

Reply:
Or this is all append-only? I was imagining the files were updated from scratch after `waitBetweenWrites`. Maybe just a periodic flush (when there have been changes) would be simpler?

Reply:
Yep, append-only - I'd initially done this async with a goroutine doing the write but realised it wasn't really needed.

I like the idea of using and resetting a ticker though, since this might end up writing more often than needed if a given write takes a long time and we then end up immediately writing again on the next iteration of the loop.

Reply:
> Would only write during step 1 and then step 5 with the combination of 2, 3 and 5 contents. Is that correct?

This is correct, but I don't think realistic for the type of data (manifest and segments) we're going to be getting from Mist.

Reply:
I realised I could achieve the same effect as resetting a ticker by just moving the point at which we reset the last write time: https://github.com/livepeer/catalyst-uploader/pull/25/commits/78de1e7efcbae53e68e93a08ac6d47b714ab535c

Reply:
LGTM!